### DISCUSSION TITLE
RFC: "registerAfterEnv" hook for "instrumentation"

### DISCUSSION THREAD

**Discussion by kettanaito** [vercel/next.js#56446] [Ideas]
### Goals

1. Allow third-party packages to patch global `fetch`. 
2. Have a manageable, framework-provided way to execute server-side side-effects in Next.js.

### Background

With the Next 13 release, Next has irreversibly removed the means to execute server-side side-effects, such as patching global `fetch` function or the `http` module. Previously, that could be achieved by executing those side-effects in `_app.js`, but the new release has no corresponding API to replace `_app` _in that regard_. 

Although this issue surfaces from MSW (https://github.com/mswjs/msw/issues/1644), it's a much broader problem since the current Next behavior doesn't allow for _any_ libraries to establish server-side side effects. 

> When I originally started the discussion around this, Next used to have _two_ development processes, which made the server-side effects introduction all the more challenging. Since then, I believe there have been [some improvements](https://github.com/mswjs/msw/issues/1644#issuecomment-1729906255), and the internal server structure has been reworked to keep just one Node.js process, which is beneficial. 

## Potential solutions

Let's discuss potential solutions to this problem using the existing state of Next, and see why they aren't fit to adequately solve it.

### Using the `instrumentation` hook

The instrumentation hook is great but its `register` function runs _before_ Next patches the global fetch, making it impossible to establish _and persist_ any patches on the function.

### Using `globalThis._nextOriginalFetch`

Next.js seems to preserve whichever state of the fetch was before the internal patching and apply the patch to that state. While this sounds like a perfect solution, it's anything but. Here are some of the disadvantages this approach has:

1. `_nextOriginalFetch` is a Next-specific _internal_ flag. No third-party consumers should depend on such implementation details, it's unsafe. 
2. Third-parties shouldn't ship a framework-specific code if those third-parties aren't framework-specific themselves; this is simply a wrong surface to tackle the problem. 

### Using testing integration in Next.js (#53409)

While this is a great endeavor, it seems to address the problem at the wrong level. 

1. The inability to establish side-effects has no direct connection to testing.
2. If talking about API mocking libraries, they can be used for far more purposes than testing on a particular level. The whole idea of tools like MSW is to elevate the network description and allow the developers to reuse mocks across the entire stack. That is possible if the tools don't lock the mocking functionality under a particular feature or use case, like it's been historically. That's kind of what this solution implies and I'm against it. 
3. The implementation of the solution also strongly suggests we are approaching the horse from the wrong end (excuse me for the metaphor here).

### Proposal

I propose to utilize the existing `instrumentation` hook. In its current form, the `register` function it supports still runs _before_ Next.js patches fetch, so whichever modifications a third-party module may establish will be discarded. 

Instead, I propose something like `registerAfterEnv` (better names are welcome), which is a new phase that Next will run _as the very last thing_ after it has prepared its environment (e.g. patched the fetch). I also take the current decision to patch `fetch` in Next as something you are choosing to stay with despite its apparent downsides and generally harmful experience for both Next users and third-party library authors shipping things for Next. 

```js
// instrumentation.js
export function registerAfterEnv() {
  console.log(fetch) // FetchPatchedByNext
  globalThis.fetch = patchFetchInAnyWay(fetch)
}
```

Perhaps the instrumentation hook being experimental is a great time to discuss this. Please, let's make this happen. Thank you.

**Comment by feedthejim** [comment]
@kettanaito Hey, thanks for writing this up. Great way of starting a productive conversation :). I also agree that adding an API to the instrumentation hook would be the best solution for your use case.

Whilst it probably won't come in the shape of `registerAfterEnv` as there's no clear boundary in Next.js of when the "env" is fully set up, I was planning on adding an `onRequest` hook that would allow you to run some logic immediately before processing a request, similarly to a middleware but for instrumentation. I believe that `fetch` would be patched at this point and that you could then add your own layer of monkey patching there :).

I don't have a clear estimate on when I would get around to doing so though.

**Reply by kettanaito** [reply]
Thanks for starting the discourse!

The `onRequest` hook you are proposing can be useful in certain situations but it won't work for what I'm advocating here. _We need a single point of entry to introduce server-side side effects **once**_. As an example, initializing MSW before every request is highly resource-inefficient and frankly wrong. 

My original thoughts on this was to propose to run the existing `register` hook _after_ Next does its `fetch` patching but I understand that may have some implications that I cannot account for. Would _that_ be a possible way forward?

**Reply by ganicus** [reply]
@feedthejim could you explain why by design there is no clear boundary when env is set up?

**Reply by feedthejim** [reply]
@kettanaito what I was implying is that you as a library user would make sure to only register it once.

```
// instrumentation.js

import { maybePatchFetch } from 'somewhere'

export function onRequest(req) {
  maybePatchFetch()
}
```

```
// somewhere.js

let patched = false;

export function maybePatchFetch() {
 if (!patched) {
    // do patching
    patched = true
  }
}
```

**Reply by kettanaito** [reply]
@feedthejim, that's understandable and we have a deduplication mechanism in place but it's simply incorrect to establish module patching before each request. This has to be done once for the entire process and that's what the `AfterEnv` part suggested, despite the otherwise unfitting name.

The `onRequest` hook locks the patching functionality even deeper into Next—something I'm encouraging with this discussion to avoid. Next is responsible for when and how it calls the `onRequest` hook, and perhaps Next will consider not calling it for certain requests in the future, breaking mocking for those requests. It's simply an improper area to establish the side effects we are talking about.

Establishing API mocking as a part of the application bootstrap is beneficial on multiple accounts. It yields a better user experience because module patching is long done once the first request is performed. Using the `onRequest` hook, the module patching will only begin before the first request.

Do you have any particular concerns about adding a hook that will be called the last thing in Next's internal bootstrapping (i.e. after it does `globalThis.fetch = ...`)?

**Reply by feedthejim** [reply]
so currently, the fetch patching occurs at request time, not when the server boots, hence why I'm suggesting to hook into `onRequest`. Another reason I'm proposing this is that this API would help some other problems. An `onNextInit` hook would be very specific for that use case, as far as I can tell, hence why I'm not too keen on adding something like that. 

Curious if you have more scenarios in mind where an AfterEnv event would be valuable.

> It yields a better user experience because module patching is long done once the first request is performed

I'm curious on how that makes it a particularly better experience. Can you expand?

> and perhaps Next will consider not calling it for certain requests in the future

Why would we?

**Reply by ganicus** [reply]
@feedthejim so is there a path or timeline toward adding more supported lifecycle hooks in next to customize as needed?

**Reply by kettanaito** [reply]
> An onNextInit hook would be very specific for that use case, as far as I can tell, hence why I'm not too keen on adding something like that.

I respect this decision. Also, thanks for explaining when exactly the fetch patching happens, it helps me a lot in assuming where such a new API may manifest better. 

> Curious if you have more scenarios in mind where an AfterEnv event would be valuable.

My main scenario and concern is that provisioning module patching _before_ any request happens leads to a better experience. The fact that Next patches fetch before request is Next's implementation detail, but it's a common practice to establish process-wide side effects, like module patching for API mocking, as early as possible in a process. 

This is likely irrelevant in the context of this discussion but one thing that makes this ASAP nature of the module patching crucial in API mocking is _import order_. There are occasions when the order when the patch is applied is important. Consider this:

```ts
import './your-code-that-makes-a-request'
import './msw-patch'
```

The way this works is the first import will keep all its globals unpatched (i.e. `http`) and the patches introduces in the second import will have no effect on the code from the first, even if requests are encapsulated in functions that are called later (those don't matter, only the imports to `http` matter).

I think we should be safe regarding this in Next but I wonder how would your proposal to have a `onRequest` hook work in scenarios like this:

```js
// app/custom/page.jsx
import makeRequest from 'some-request-lib'

// This is a terrible pseudo code.
export async function GET() {
  const res = await makeRequest()
}
```

When will the `onRequest` hook execute in relation to the import of `some-request-lib`? 

**It's crucial that the new hook executes before this entire module**. Otherwise, the same import order issue I mentioned above will occur. Next users that rely on global fetch may be fine but Next users that import third-party request libraries may be subjected to import order issue, breaking their expectations regarding API mocking. 

## Naming

I suggest renaming `onRequest` to `beforeRequest` to emphasize _when_ this hook executes. I believe it becomes more apparent with that phrasing. Curious to hear what you think.

**Reply by feedthejim** [reply]
What you're saying makes a lot of sense @kettanaito, I think the patching could in theory done earlier.

The example you provided is a good argument for making the fetch patch happen earlier as AFAIK, currently, this kind of code would bypass the fetch cache on the first request.

> I suggest renaming onRequest to beforeRequest to emphasize when this hook executes. I believe it becomes more apparent with that phrasing. Curious to hear what you think.

still thinking about the final API, I think the function might need to be able to observe the execution of the request so either names would not be particularly appropriate.

**Reply by kettanaito** [reply]
Happy to have explained my point well! From Next's perspective, the only thing that matters is that Next patches `fetch` before the Next user calls it. It's a self-contained sequence. But tools like MSW don't really control when the requests may happen (as I've illustrated with the import order example), and this makes them sensitive as to when you initiate them. 

@feedthejim, what are the implications of utilizing the `register` hook to solve this by making it run _after_ Next patches fetch? Is that entirely impossible due to when Next patches `fetch`?

> I'm quite curious what's the background behind the decision to patch `fetch` on an individual request basis if the patch is always the same? Next can also benefit from early patching, although quite less significantly than third parties. 

From what I understood, we are a bit in a pickle now with these statements:

1. API mocking must initiate itself as early as possible;
2. Next patches `fetch` before each request.

The current Next behavior will throw away whichever module patches will be introduced even in the newly discussed API due to the request-based fetch patching. 

May I propose to reconsider the decision to provision fetch patching on the request basis? This kind of change can be tackled differently based on the history of it:

- Optimistically, move the fetch patch in Next to sometime before/after "register" (i.e. decouple it from the request).
- If there are any requirements to have access to the individual request within the fetch patching logic, those requirements can be met by either:
  - Accessing the request instance in the `fetch` patch itself—it's there.
  - Splitting this fetch patching in Next into two chunks: the first applies on the environment level, the second applies on the request level. I still think this splitting would be inefficient. I can't think of a use case where patching fetch before each request would be beneficial but I trust you know more about this.

**Reply by frodri** [reply]
@feedthejim Have you had a chance to review the questions provided by @kettanaito last month?

**Reply by kettanaito** [reply]
@frodri, I actually have a bit of news. Lee has [shared with me](https://twitter.com/leeerob/status/1720811544760438864) on Twitter that they are moving away from patching `fetch` altogether. I believe that takes a bigger priority for them than this proposal, especially since once they stop patching `fetch` this proposal would be rendered redundant. 

I would still like an official statement from someone in the Next.js team regarding what's the way forward here. A lot of people are blocked by their decisions and cannot continue developing and testing their Next.js apps with MSW until that decision is made.

**Reply by jviall** [reply]
> they are moving away from patching fetch altogether

@kettanaito If this is the case, does that mean that attempting a `fetch` patch in the `register` hook would eventually "just work"?

**Reply by kettanaito** [reply]
@jviall, I'd expect it to work, yes. Since nothing else will override `fetch` after the `register` instrumentation hook. I don't recall if that hook runs once or not but that shouldn't be a problem.

**Reply by edmolima** [reply]
Hey @jviall @kettanaito Did you guys deal with it? I mean, Is there some alternative that you are using until somehow the vercel team responds about this officially proposed RFC?

**Reply by kettanaito** [reply]
@edmolima, some folks are moving the MSW request handlers into a standalone server using `@mswjs/http-middleware` (https://github.com/mswjs/msw/issues/1644#issuecomment-1750722052) and then routing the app's traffic through it. This will work since it negates the need to intercept anything but keep in mind this approach negates the main benefits of MSW. 

There are no other workarounds here because of how Next.js behaves at the moment.

**Reply by kmiwa007** [reply]
Just chiming in support that Next needs make MSW happen!

**Reply by rkofman** [reply]
@feedthejim (or others at Vercel): is there by any chance a rough timeline for when a version of Next.js will be released that doesn't override `fetch`, and therefore works well with `msw` (and others)?

**Reply by rmiller61** [reply]
Would also like to second this. The inability to easily mock APIs has become a major wedge issue for considering Next in my projects.

**Comment by Geminii** [comment]
@feedthejim Any update on this topic ? 🙏 
Christmas before Christmas ? 😍

**Reply by kettanaito** [reply]
I don't think it's reasonable to expect the team to release an experimental Cache API before the holidays. Let people rest. I've seen some code snippets of the experimental Cache API in the wild, so I assume it's there and the effort is ongoing. In a few months, perhaps?

**Reply by Geminii** [reply]
Glad to heard, thanks for quick update 🙂
Happy holidays and see you next year 🎄

**Comment by AndrewIsh** [comment]
Is progress on this still being tracked here? Is there any progress? :)

**Comment by SalahAdDin** [comment]
Another important issue being left part by Vercel, is it the time to move to another framework?

**Reply by chrisb2244** [reply]
With respect to the issue described specifically for fetch, a recent(ish) patch (maybe a month or a month and a half ago) changed Next's behaviour here - patched fetch is preserved now (and you can use MSW for server-side code).

It's still a little confusing given your nextjs code can run in various places and if you want to intercept them all, you'll need a combination of worker/server MSW setups, but you can of course reuse the handlers.

**Reply by rmiller61** [reply]
> With respect to the issue described specifically for fetch, a recent(ish) patch (maybe a month or a month and a half ago) changed Next's behaviour here - patched fetch is preserved now (and you can use MSW for server-side code).
> 
> It's still a little confusing given your nextjs code can run in various places and if you want to intercept them all, you'll need a combination of worker/server MSW setups, but you can of course reuse the handlers.

Do you have a working example? AFAIK the cache APIs aren't stable.

**Reply by chrisb2244** [reply]
If the goal is only to provide interception with static handlers, then I believe Artem's example here is perhaps working?
https://github.com/mswjs/examples/pull/101

I have an example repo that I started on here (https://github.com/chrisb2244/msw-nextjs-example) but I'm currently waiting to see if a PR gets merged (https://github.com/mswjs/msw/pull/2041) before I return to it (if I recall correctly, I was fairly happy with the server-side fetches and the remote server control for testing (working together with static handlers, if desired), but the client-side stuff left something to be desired (probably because it's been too long since I tried to put Playwright and MSW together, and I've forgotten the key points)).

I'll be going back to that example if I can in the future to try and test out various behaviours, in part for my own sanity, testing in a simple project before trying to apply the same to a real repository (where I'd be less sure that any failures weren't just my code being exercised in ways I hadn't expected... :/ )

**Reply by SalahAdDin** [reply]
@chrisb2244 

> If the goal is only to provide interception with static handlers, then I believe Artem's example here is perhaps working.
> https://github.com/mswjs/examples/pull/101


No, it does not work. MSW is unable to intercept the request.
